python-checks/src/main/java/org/sonar/python/checks/hotspots/GraphQLIntrospectionCheck.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks.hotspots;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.ExpressionList;
+import org.sonar.plugins.python.api.tree.ListLiteral;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tuple;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6786")
+public class GraphQLIntrospectionCheck extends PythonSubscriptionCheck {
+
+  private static final Set<String> GRAPHQL_VIEWS_FQNS = Set.of(
+    "flask_graphql.GraphQLView.as_view",
+    "graphql_server.flask.GraphQLView.as_view");
+
+  private static final Set<String> SENSITIVE_VALIDATION_RULE_FQNS = Set.of(
+    "graphene.validation.DisableIntrospection",
+    "graphql.validation.NoSchemaIntrospectionCustomRule");
+
+  private static final String MESSAGE = "Disable introspection on this GraphQL server endpoint.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, GraphQLIntrospectionCheck::checkGraphQLIntrospection);
+  }
+
+  private static void checkGraphQLIntrospection(SubscriptionContext ctx) {
+    CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+    Optional.ofNullable(callExpression.calleeSymbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter(fqn -> GRAPHQL_VIEWS_FQNS.contains(fqn) &&
```

Comment:
I guess it could be split up by two `filter` method calls:
```java
.filter(GRAPHQL_VIEWS_FQNS::contains)
.filter(fqn -> hasSensitiveMiddlewares(callExpression.arguments()) || hasSensitiveValidationRules(callExpression.arguments())
```
or since you are working with the `callExpression` all the time - maybe we don't need to extract the symbol at all:
```java
Optional.of(callExpression)
  .filter(exp -> Optional.of(exp)
    .map(CallExpression::calleeSymbol)
    .map(Symbol::fullyQualifiedName)
    .stream()
    .allMatch(GRAPHQL_VIEWS_FQNS::contains)
  )
  .filter(exp -> hasSensitiveMiddlewares(callExpression.arguments()) || hasSensitiveValidationRules(callExpression.arguments()))
  .ifPresent(exp -> ctx.addIssue(exp, MESSAGE)
```

Reply:
The two filters make it clearer, thank you. I will keep the non-nested version though. I think the intent is still without the call expression being passed all the way down